docs/design/1-support-task-progress.md

```diff
@@ -0,0 +1,85 @@
+---
+author: Lance Ren <lanceren@yunify.com>
+status: draft
+updated_at: 2020-02-19
+---
+
+# Proposal: Support Task Progress
+
+## Background
+
+For many of our tasks, they cost a lot while running. In case of improving
+user's experience, we do need to support return task progress state.
+The task caller may use the state for implementing the process bar and so on.
+
+## Proposal
+
+So I propose following changes:
+
+- Add a type (maybe like: progress by percent) in time-cost task,
+  so that we can get progress now.
+- Set progress state asynchronously in the progress of task running.
+
+After all these work, we can work well with progress bar:
+
+```go
+package main
+
+import (
+    "log"
+    "time"
+
+	"github.com/qingstor/noah/task"
+    "github.com/schollz/progressbar/v2"
+)
+
+func main()  {
```

Comment:
From my point of view, noah's task should:

- allow us pass a callback or something and noah will update it regularly
- allow us call `Progress() float` or something so we can get current state

Maybe we should take them both into consideration?